From 3d9eaea426c5d54c9af76ca7201386b42f021906 Mon Sep 17 00:00:00 2001
From: olegsvs <oleg.texet@gmail.com>
Date: Fri, 24 Jun 2016 14:46:43 -0400
Subject: [PATCH] Battery

---
 healthd/BatteryMonitor.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/healthd/BatteryMonitor.cpp b/healthd/BatteryMonitor.cpp
index 66c4e8e..e5c3df8 100644
--- a/healthd/BatteryMonitor.cpp
+++ b/healthd/BatteryMonitor.cpp
@@ -198,7 +198,7 @@ bool BatteryMonitor::update(void) {
     props.batteryLevel = mBatteryFixedCapacity ?
         mBatteryFixedCapacity :
         getIntField(mHealthdConfig->batteryCapacityPath);
-    props.batteryVoltage = getIntField(mHealthdConfig->batteryVoltagePath) / 1000;
+    props.batteryVoltage = getIntField(mHealthdConfig->batteryVoltagePath) ;
 
     props.batteryTemperature = mBatteryFixedTemperature ?
         mBatteryFixedTemperature :
-- 
1.9.1

